src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -978,6 +978,31 @@ static BOOL AttributeHasChangesOmittedQuality(MTRAttributePath * attributePath)
     return attributeValueToReturn;
 }
 
+- (nullable id)readAttributeWithEndpointID:(NSNumber *)endpointID
+                                 clusterID:(NSNumber *)clusterID
+                               attributeID:(NSNumber *)attributeID
+                                    params:(MTRReadParams * _Nullable)params
+                                     error:(NSError * __autoreleasing *)error
+{
+    NSDictionary<NSString *, id> * value = [self readAttributeWithEndpointID:endpointID
+                                                                   clusterID:clusterID
+                                                                 attributeID:attributeID
+                                                                      params:params];
+    if (value == nil) {
+        return nil;
+    }
+
+    auto * path = [MTRAttributePath attributePathWithEndpointID:endpointID clusterID:clusterID attributeID:attributeID];
+
+    auto * report = [[MTRAttributeReport alloc] initWithResponseValue:@{
+        MTRAttributePathKey : path,
+        MTRDataKey : value,
+    }
+                                                                error:error];
+
+    return report.value;
```

Comment:
Don't you have to handle the case of `report.error` being populated instead of `report.value` ?

Reply:
`report.error` is never going to be populated here, given the API documentation for `initWithResponseValue` and its implementation.  However it _can_ return nil for `report`.  I should probably handle that more explicitly, instead of relying on `report.value` being nil if `report` is nil.

Reply:
I'd say that relying on `report.value` being nil if report is nil is idiomatic in ObjC.

I did look at the docs for initWithResponseValue which say
```
 * Will set the value property to nil and the error property to non-nil, even if
 * the schema for the value is not known, if the response-value is an error, not
 * data.
```
It wasn't clear to me that that's not a possibility here.

Reply:
Ah, an error response value would not have `MTRDataKey` in the dictionary.  It would have `MTRErrorKey`.

Reply:
And more to the point, the attribute cache in MTRDevice only stores attribute values, and with the path stripped off (which is why we have to add the path here).  For errors, it reports the error to the listener, but does not store it.

Reply:
Documented things a bit better.